### PR TITLE
Fix parent schema discriminator validation

### LINKED ISSUE #402: Schema array type with anyOf and discriminator fails validation
I'm trying to validate an OpenAPI document that contains a schema object that is of type 'array' that can be an 'anyOf' that uses a discriminator to know which object goes into the array and it's failing validation.  The document appears to be a valid OAS3 document and the Swagger UI has no issues rendering it, so I'm guessing it's an issue with OpenAPI.NET.

I'm getting validation errors similar to:

> Composite Schema ChildTypeA must contain property specified in the discriminator type. [#/components/schemas/SampleResponse/properties/suggestedActions/items/anyOf]
Composite schema ChildTypeA must contain property specified in the discriminator type in the required field list. [#/components/schemas/SampleResponse/properties/suggestedActions/items/anyOf]

...for this document

```
openapi: 3.0.0
info:
  description: Service to test discriminator
  version: 1.0.0
  title: Test Service
servers:
  - url: 'https://localhost'
paths:
  /test:
    post:
      description: Sample post of SampleRequest
      requestBody:
        description: Returns the SampleResponse with multiple types
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/SampleRequest'
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/SampleResponse'
components:
  schemas:
    SampleRequest:
      type: object
      required:
        - id
      properties:
        id:
          type: string
          description: Random ID
          example: '1513776099991'
    BaseType:
      type: object
      required:
        - type
      properties:
        type:
          type: string
          description: Suggested object type
    ChildTypeA:
      allOf:
        - $ref: '#/components/schemas/BaseType'
          properties:
            type:
              type: string
              enum:
                - typeA
      type: object
    ChildTypeB:
      allOf:
        - $ref: '#/components/schemas/BaseType'
          properties:
            type:
              type: string
              enum:
                - typeB
      type: object
    SampleResponse:
      type: object
      properties:
        suggestedActions:
          type: array
          items:
            anyOf:
              - $ref: '#/components/schemas/ChildTypeA'
              - $ref: '#/components/schemas/ChildTypeB'
            discriminator:
              propertyName: type
```

### PR DESCRIPTION
**Pull request by MaggieKimani1**
Fixes #402

### CODE REVIEW

**Comment by darrelmiller** [comment]
The validation process needs to be recursive.  The current check only validates direct children of the schema.  The following example still fails.

```yaml
openapi: 3.0.0
info:
  description: Service to test discriminator
  version: 1.0.0
  title: Test Service
servers:
  - url: 'https://localhost'
paths:
  /test:
    post:
      description: Sample post
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Parent'
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Parent'
components:
  schemas:
    Parent:
      type: object
      oneOf:
         - $ref: '#/components/schemas/ChildA'
         - $ref: '#/components/schemas/ChildB'
      discriminator:
         propertyName: type
    ChildA:
      allOf:
      - $ref: "#/components/schemas/BaseChild"
      - properties:
          AName:
            type: string
        type: object
    ChildB:
      allOf:
      - $ref: "#/components/schemas/BaseChild"
      - properties:
          BName:
           type: string
        type: object
    BaseChild:
      properties:
        type:
         type: string
      type: object

```

**Review by darrelmiller — CHANGES_REQUESTED** [review]
The checking needs to be made recursive.  See the example I included in the overview comments.

**Comment by MaggieKimani1** [comment]
> The checking needs to be made recursive. See the example I included in the overview comments.

Thanks for the correction. Resolving..

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 1]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+
+        /// <summary>
+        /// Validates the property name in the discriminator against the ones present in the children schema
+        /// </summary>
+        /// <param name="schema">The parent schema.</param>
+        /// <param name="discriminator">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="schemaReferenceId"></param>
+        /// <param name="context">A validation context.</param>
+        public static bool ValidateChildSchemaAgainstDiscriminator(OpenApiSchema schema, string discriminator, string schemaReferenceId, IValidationContext context)
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:93]:
```suggestion
        public static bool ValidateChildSchemaAgainstDiscriminator(OpenApiSchema schema, string discriminatorName)
```

- I don't believe you require the other params.
- You can also consider renaming `discriminator` to `discriminatorName` to convey the proper intent.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 2]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+        /// Traverses the schema elements and checks whether the schema contains the discriminator.
+        /// </summary>
+        /// <param name="discriminator">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="childSchema">The child schema.</param>
+        /// <param name="schemaReferenceId"> The schema reference Id.</param>
+        /// <param name="context"> A validation context.</param>
+        /// <param name="containsDiscriminator">Tracks whether the discriminator is present.</param>
+        /// <returns></returns>
+        public static bool TraverseSchemaElements(string discriminator, IList<OpenApiSchema> childSchema, string schemaReferenceId, IValidationContext context, bool containsDiscriminator)
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:127]:
```suggestion
        public static bool TraverseSchemaElements(string discriminatorName, IList<OpenApiSchema> childSchema, ref bool containsDiscriminator)
```

`containsDiscriminator` is a value type, you may need to [pass by reference](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref#passing-an-argument-by-reference), i.e. `ref` if you want its value to be modified by the called method.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 3]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+        {
+            foreach (var childItem in childSchema)
+            {
+                if (!childItem.Properties.ContainsKey(discriminator) && !childItem.Required.Contains(discriminator))
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:131]:
```suggestion
                if ((!childItem.Properties?.ContainsKey(discriminatorName) ?? false) &&
                    (!childItem.Required?.Contains(discriminatorName) ?? false))
```
Guard against potential null references.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 4]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+        /// <param name="schema">The parent schema.</param>
+        /// <param name="discriminator">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="schemaReferenceId"></param>
+        /// <param name="context">A validation context.</param>
+        public static bool ValidateChildSchemaAgainstDiscriminator(OpenApiSchema schema, string discriminator, string schemaReferenceId, IValidationContext context)
+        {
+            bool containsDiscriminator = false;
+
+            if (!schema.Required.Contains(discriminator))
```
Comment (irvinesunday) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:97]:
```suggestion
            if (!schema.Required?.Contains(discriminatorName) ?? false)
```

Guard against potential null references.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs (hunk 1 above)
Comment (MaggieKimani1) [reply src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:93]: Good point.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 5]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+
+        /// <summary>
+        /// Validates the property name in the discriminator against the ones present in the children schema
+        /// </summary>
+        /// <param name="schema">The parent schema.</param>
+        /// <param name="discriminatorName">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        public static bool ValidateChildSchemaAgainstDiscriminator(OpenApiSchema schema, string discriminatorName)
+        {
+            bool containsDiscriminator = false;
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:92]:
```suggestion
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 6]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+
+            if (!schema.Required?.Contains(discriminatorName) ?? false)
+            {
+                // recursively check nested schema.OneOf, schema.AnyOf or schema.AllOf and their required fields for the discriminator
+                if (schema.OneOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminatorName, schema.OneOf, ref containsDiscriminator);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:99]:
```suggestion
                    return TraverseSchemaElements(discriminatorName, schema.OneOf);
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 7]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+                }
+                if (schema.AnyOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminatorName, schema.AnyOf, ref containsDiscriminator);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:103]:
```suggestion
                    return TraverseSchemaElements(discriminatorName, schema.AnyOf);
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 8]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+                }
+                if (schema.AllOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminatorName, schema.AllOf, ref containsDiscriminator);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:107]:
```suggestion
                    return TraverseSchemaElements(discriminatorName, schema.AllOf);
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 9]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+                }
+            }
+            else
+            {
+                return true;
+            }
+
+            return containsDiscriminator;
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:115]:
```suggestion
            return false;
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 10]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+
+        /// <summary>
+        /// Traverses the schema elements and checks whether the schema contains the discriminator.
+        /// </summary>
+        /// <param name="discriminatorName">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="childSchema">The child schema.</param>
+        /// <param name="containsDiscriminator">Tracks whether the discriminator is present.</param>
+        /// <returns></returns>
+        public static bool TraverseSchemaElements(string discriminatorName, IList<OpenApiSchema> childSchema, ref bool containsDiscriminator)
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:126]:
```suggestion
        public static bool TraverseSchemaElements(string discriminatorName, IList<OpenApiSchema> childSchema)
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 11]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+            return containsDiscriminator;
+        }
+
+        /// <summary>
+        /// Traverses the schema elements and checks whether the schema contains the discriminator.
+        /// </summary>
+        /// <param name="discriminatorName">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="childSchema">The child schema.</param>
+        /// <param name="containsDiscriminator">Tracks whether the discriminator is present.</param>
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:124]:
```suggestion
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 12]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+            foreach (var childItem in childSchema)
+            {
+                if ((!childItem.Properties?.ContainsKey(discriminatorName) ?? false) &&
+                                    (!childItem.Required?.Contains(discriminatorName) ?? false))
+                {
+                    return ValidateChildSchemaAgainstDiscriminator(childItem, discriminatorName);
+                }
+                else
+                {
+                    return containsDiscriminator = true;
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:137]:
```suggestion
                    return containsDiscriminator;
```

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 13]
```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
...
+                }
+            }
+
+            return containsDiscriminator;
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs:141]:
```suggestion
            return false;
```

**Review by baywet — CHANGES_REQUESTED** [review]

**Review by baywet — APPROVED** [review]

**Review by darrelmiller — APPROVED** [review]

**Pull request merged**